File: frameworks/python/src/dpe/_accumulators.py
"""Typed accumulators for shared memory across processors."""

from collections import deque


class Number:
    """Numeric accumulator with increment/decrement."""

    __slots__ = ("_value",)

    def __init__(self, initial: float = 0):
        self._value = initial

    def inc(self, delta: float = 1) -> float:
        self._value += delta
        return self._value

    def dec(self, delta: float = 1) -> float:
        self._value -= delta
        return self._value

    @property
    def value(self) -> float:
        return self._value


class Average:
    """Running average without storing all values."""

    __slots__ = ("_sum", "_count")

    def __init__(self):
        self._sum = 0.0
        self._count = 0

    def add(self, value: float):
        self._sum += value
        self._count += 1

    @property
    def value(self) -> float:
        if self._count == 0:
            return 0.0
        return self._sum / self._count

    @property
    def count(self) -> int:
        return self._count


class MinMax:
    """Track minimum and maximum values."""

    __slots__ = ("_min", "_max", "_count")

    def __init__(self):
        self._min = None
        self._max = None
        self._count = 0

    def add(self, value: float):
        if self._min is None or value < self._min:
            self._min = value
        if self._max is None or value > self._max:
            self._max = value
        self._count += 1

    @property
    def min(self) -> float | None:
        return self._min

    @property
    def max(self) -> float | None:
        return self._max

    @property
    def count(self) -> int:
        return self._count


class Set:
    """Unique value collection."""

    __slots__ = ("_data",)

    def __init__(self):
        self._data = set()

    def add(self, value):
        self._data.add(value)

    def has(self, value) -> bool:
        return value in self._data

    def delete(self, value):
        self._data.discard(value)

    @property
    def size(self) -> int:
        return len(self._data)

    @property
    def values(self) -> list:
        return list(self._data)


class Map:
    """Key-value lookup. Transient storage, not for accumulation."""

    __slots__ = ("_data",)

    def __init__(self):
        self._data = {}

    def set(self, key: str, value):
        self._data[key] = value

    def get(self, key: str, default=None):
        return self._data.get(key, default)

    def has(self, key: str) -> bool:
        return key in self._data

    def delete(self, key: str):
        self._data.pop(key, None)

    @property
    def keys(self) -> list:
        return list(self._data.keys())

    @property
    def size(self) -> int:
        return len(self._data)


class Buffer:
    """Ordered collection with optional size limit (sliding window)."""

    __slots__ = ("_data", "_limit")

    def __init__(self, limit: int | None = None):
        self._data: deque = deque(maxlen=limit)
        self._limit = limit

    def push(self, item):
        self._data.append(item)

    @property
    def items(self) -> list:
        return list(self._data)

    @property
    def size(self) -> int:
        return len(self._data)

    def clear(self):
        self._data.clear()


class Boolean:
    """Simple flag."""

    __slots__ = ("_value",)

    def __init__(self, initial: bool = False):
        self._value = initial

    def set(self, value: bool = True):
        self._value = value

    def toggle(self, value: bool | None = None):
        if value is not None:
            self._value = value
        else:
            self._value = not self._value

    @property
    def value(self) -> bool:
        return self._value


class BitMask:
    """Named flag set for tracking multiple boolean states."""

    __slots__ = ("_flags",)

    def __init__(self, names: list[str]):
        self._flags = {name: False for name in names}

    def set(self, name: str):
        if name not in self._flags:
            raise KeyError(f"Unknown flag: {name}")
        self._flags[name] = True

    def get(self, name: str) -> bool:
        if name not in self._flags:
            raise KeyError(f"Unknown flag: {name}")
        return self._flags[name]

    def clear(self, name: str):
        if name not in self._flags:
            raise KeyError(f"Unknown flag: {name}")
        self._flags[name] = False

    def all(self) -> bool:
        return all(self._flags.values())

    def any(self) -> bool:
        return any(self._flags.values())


class Trigger:
    """Condition checker across accumulators. Condition is a callable(memory) → bool."""

    __slots__ = ("_condition", "_memory_ref")

    def __init__(self, condition, memory):
        self._condition = condition
        self._memory_ref = memory

    def check(self) -> bool:
        return self._condition(self._memory_ref)


class Memory:
    """Shared memory — typed accumulator registry.

    Accumulators are created on first access by name.
    Same name always returns the same instance.
    """

    def __init__(self):
        self._store: dict = {}

    def number(self, name: str, initial: float = 0) -> Number:
        return self._get_or_create(name, Number, initial)

    def average(self, name: str) -> Average:
        return self._get_or_create(name, Average)

    def minmax(self, name: str) -> MinMax:
        return self._get_or_create(name, MinMax)

    def set(self, name: str) -> Set:
        return self._get_or_create(name, Set)

    def map(self, name: str) -> Map:
        return self._get_or_create(name, Map)

    def buffer(self, name: str, limit: int | None = None) -> Buffer:
        return self._get_or_create(name, Buffer, limit)

    def boolean(self, name: str, initial: bool = False) -> Boolean:
        return self._get_or_create(name, Boolean, initial)

    def bitmask(self, name: str, flags: list[str]) -> BitMask:
        return self._get_or_create(name, BitMask, flags)

    def trigger(self, name: str, condition=None) -> Trigger:
        if name in self._store:
            return self._store[name]
        if condition is None:
            raise ValueError(f"Trigger '{name}' does not exist and no condition provided")
        t = Trigger(condition, self)
        self._store[name] = t
        return t

    def get(self, name: str):
        """Get accumulator by name. Returns None if not found."""
        return self._store.get(name)

    def delete(self, name: str):
        """Remove accumulator by name."""
        self._store.pop(name, None)

    def has(self, name: str) -> bool:
        return name in self._store

    def _get_or_create(self, name, cls, *args):
        existing = self._store.get(name)
        if existing is not None:
            if not isinstance(existing, cls):
                raise TypeError(
                    f"Accumulator '{name}' exists as {type(existing).__name__}, "
                    f"not {cls.__name__}"
                )
            return existing
        instance = cls(*args)
        self._store[name] = instance
        return instance

File: frameworks/rust/src/accumulators.rs
//! Typed accumulators for shared memory across processors.

use std::collections::{HashMap, HashSet, VecDeque};
use serde_json::Value;

/// Numeric accumulator with increment/decrement.
pub struct Number {
    value: f64,
}

impl Number {
    pub fn new(initial: f64) -> Self {
        Self { value: initial }
    }

    pub fn inc(&mut self, delta: f64) -> f64 {
        self.value += delta;
        self.value
    }

    pub fn dec(&mut self, delta: f64) -> f64 {
        self.value -= delta;
        self.value
    }

    pub fn value(&self) -> f64 {
        self.value
    }
}

/// Running average without storing all values.
#[derive(Default)]
pub struct Average {
    sum: f64,
    count: u64,
}

impl Average {
    pub fn new() -> Self {
        Self { sum: 0.0, count: 0 }
    }

    pub fn add(&mut self, value: f64) {
        self.sum += value;
        self.count += 1;
    }

    pub fn value(&self) -> f64 {
        if self.count == 0 {
            0.0
        } else {
            self.sum / self.count as f64
        }
    }

    pub fn count(&self) -> u64 {
        self.count
    }
}

/// Track minimum and maximum values.
#[derive(Default)]
pub struct MinMax {
    min: Option<f64>,
    max: Option<f64>,
    count: u64,
}

impl MinMax {
    pub fn new() -> Self {
        Self { min: None, max: None, count: 0 }
    }

    pub fn add(&mut self, value: f64) {
        self.min = Some(self.min.map_or(value, |m| m.min(value)));
        self.max = Some(self.max.map_or(value, |m| m.max(value)));
        self.count += 1;
    }

    pub fn min(&self) -> Option<f64> {
        self.min
    }

    pub fn max(&self) -> Option<f64> {
        self.max
    }

    pub fn count(&self) -> u64 {
        self.count
    }
}

/// Unique value collection.
#[derive(Default)]
pub struct Set {
    data: HashSet<String>,
}

impl Set {
    pub fn new() -> Self {
        Self { data: HashSet::new() }
    }

    pub fn add(&mut self, value: &str) {
        self.data.insert(value.to_string());
    }

    pub fn has(&self, value: &str) -> bool {
        self.data.contains(value)
    }

    pub fn delete(&mut self, value: &str) {
        self.data.remove(value);
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }

    pub fn values(&self) -> Vec<&String> {
        self.data.iter().collect()
    }
}

/// Key-value lookup. Transient storage.
#[derive(Default)]
pub struct Map {
    data: HashMap<String, Value>,
}

impl Map {
    pub fn new() -> Self {
        Self { data: HashMap::new() }
    }

    pub fn set(&mut self, key: &str, value: Value) {
        self.data.insert(key.to_string(), value);
    }

    pub fn get(&self, key: &str) -> Option<&Value> {
        self.data.get(key)
    }

    pub fn has(&self, key: &str) -> bool {
        self.data.contains_key(key)
    }

    pub fn delete(&mut self, key: &str) {
        self.data.remove(key);
    }

    pub fn keys(&self) -> Vec<&String> {
        self.data.keys().collect()
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }
}

/// Ordered collection with optional size limit (sliding window).
pub struct Buffer {
    data: VecDeque<Value>,
    limit: Option<usize>,
}

impl Buffer {
    pub fn new(limit: Option<usize>) -> Self {
        Self {
            data: VecDeque::new(),
            limit,
        }
    }

    pub fn push(&mut self, item: Value) {
        if let Some(limit) = self.limit {
            if self.data.len() >= limit {
                self.data.pop_front();
            }
        }
        self.data.push_back(item);
    }

    pub fn items(&self) -> Vec<&Value> {
        self.data.iter().collect()
    }

    pub fn size(&self) -> usize {
        self.data.len()
    }

    pub fn clear(&mut self) {
        self.data.clear();
    }
}

/// Simple flag.
pub struct Boolean {
    value: bool,
}

impl Boolean {
    pub fn new(initial: bool) -> Self {
        Self { value: initial }
    }

    pub fn set(&mut self, value: bool) {
        self.value = value;
    }

    pub fn toggle(&mut self, value: Option<bool>) {
        match value {
            Some(v) => self.value = v,
            None => self.value = !self.value,
        }
    }

    pub fn value(&self) -> bool {
        self.value
    }
}

/// Named flag set for tracking multiple boolean states.
pub struct BitMask {
    flags: HashMap<String, bool>,
}

impl BitMask {
    pub fn new(names: &[&str]) -> Self {
        let flags = names.iter().map(|n| (n.to_string(), false)).collect();
        Self { flags }
    }

    pub fn set(&mut self, name: &str) -> Result<(), String> {
        match self.flags.get_mut(name) {
            Some(v) => { *v = true; Ok(()) }
            None => Err(format!("Unknown flag: {}", name)),
        }
    }

    pub fn get(&self, name: &str) -> Result<bool, String> {
        self.flags.get(name).copied().ok_or_else(|| format!("Unknown flag: {}", name))
    }

    pub fn clear(&mut self, name: &str) -> Result<(), String> {
        match self.flags.get_mut(name) {
            Some(v) => { *v = false; Ok(()) }
            None => Err(format!("Unknown flag: {}", name)),
        }
    }

    pub fn all(&self) -> bool {
        self.flags.values().all(|&v| v)
    }

    pub fn any(&self) -> bool {
        self.flags.values().any(|&v| v)
    }
}

/// Condition checker. Condition is a closure over Memory.
pub struct Trigger {
    condition: Box<dyn Fn(&Memory) -> bool>,
}

impl Trigger {
    pub fn new(condition: impl Fn(&Memory) -> bool + 'static) -> Self {
        Self { condition: Box::new(condition) }
    }

    pub fn check(&self, memory: &Memory) -> bool {
        (self.condition)(memory)
    }
}

/// Accumulator enum for type-safe storage in Memory.
pub enum Accumulator {
    Number(Number),
    Average(Average),
    MinMax(MinMax),
    Set(Set),
    Map(Map),
    Buffer(Buffer),
    Boolean(Boolean),
    BitMask(BitMask),
    Trigger(Trigger),
}

/// Shared memory — typed accumulator registry.
#[derive(Default)]
pub struct Memory {
    store: HashMap<String, Accumulator>,
}

impl Memory {
    pub fn new() -> Self {
        Self { store: HashMap::new() }
    }

    pub fn number(&mut self, name: &str, initial: f64) -> &mut Number {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Number(Number::new(initial)));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Number(n) => n,
            _ => panic!("Accumulator '{}' is not a Number", name),
        }
    }

    pub fn average(&mut self, name: &str) -> &mut Average {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Average(Average::new()));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Average(a) => a,
            _ => panic!("Accumulator '{}' is not an Average", name),
        }
    }

    pub fn minmax(&mut self, name: &str) -> &mut MinMax {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::MinMax(MinMax::new()));
        match self.store.get_mut(name).unwrap() {
            Accumulator::MinMax(m) => m,
            _ => panic!("Accumulator '{}' is not a MinMax", name),
        }
    }

    pub fn set(&mut self, name: &str) -> &mut Set {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Set(Set::new()));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Set(s) => s,
            _ => panic!("Accumulator '{}' is not a Set", name),
        }
    }

    pub fn map(&mut self, name: &str) -> &mut Map {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Map(Map::new()));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Map(m) => m,
            _ => panic!("Accumulator '{}' is not a Map", name),
        }
    }

    pub fn buffer(&mut self, name: &str, limit: Option<usize>) -> &mut Buffer {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Buffer(Buffer::new(limit)));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Buffer(b) => b,
            _ => panic!("Accumulator '{}' is not a Buffer", name),
        }
    }

    pub fn boolean(&mut self, name: &str, initial: bool) -> &mut Boolean {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Boolean(Boolean::new(initial)));
        match self.store.get_mut(name).unwrap() {
            Accumulator::Boolean(b) => b,
            _ => panic!("Accumulator '{}' is not a Boolean", name),
        }
    }

    pub fn bitmask(&mut self, name: &str, flags: &[&str]) -> &mut BitMask {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::BitMask(BitMask::new(flags)));
        match self.store.get_mut(name).unwrap() {
            Accumulator::BitMask(b) => b,
            _ => panic!("Accumulator '{}' is not a BitMask", name),
        }
    }

    pub fn trigger(&mut self, name: &str, condition: impl Fn(&Memory) -> bool + 'static) {
        self.store.entry(name.to_string())
            .or_insert_with(|| Accumulator::Trigger(Trigger::new(condition)));
    }

    pub fn check_trigger(&self, name: &str) -> bool {
        match self.store.get(name) {
            Some(Accumulator::Trigger(t)) => t.check(self),
            _ => false,
        }
    }

    pub fn has(&self, name: &str) -> bool {
        self.store.contains_key(name)
    }

    pub fn delete(&mut self, name: &str) {
        self.store.remove(name);
    }

    // -- Immutable getters for reading accumulator values (used in trigger closures) --

    pub fn get_number(&self, name: &str) -> Option<&Number> {
        match self.store.get(name) {
            Some(Accumulator::Number(n)) => Some(n),
            _ => None,
        }
    }

    pub fn get_average(&self, name: &str) -> Option<&Average> {
        match self.store.get(name) {
            Some(Accumulator::Average(a)) => Some(a),
            _ => None,
        }
    }

    pub fn get_minmax(&self, name: &str) -> Option<&MinMax> {
        match self.store.get(name) {
            Some(Accumulator::MinMax(m)) => Some(m),
            _ => None,
        }
    }

    pub fn get_set(&self, name: &str) -> Option<&Set> {
        match self.store.get(name) {
            Some(Accumulator::Set(s)) => Some(s),
            _ => None,
        }
    }

    pub fn get_map(&self, name: &str) -> Option<&Map> {
        match self.store.get(name) {
            Some(Accumulator::Map(m)) => Some(m),
            _ => None,
        }
    }

    pub fn get_boolean(&self, name: &str) -> Option<&Boolean> {
        match self.store.get(name) {
            Some(Accumulator::Boolean(b)) => Some(b),
            _ => None,
        }
    }

    pub fn get_bitmask(&self, name: &str) -> Option<&BitMask> {
        match self.store.get(name) {
            Some(Accumulator::BitMask(b)) => Some(b),
            _ => None,
        }
    }
}
